src/models/preprocessor2.py

```diff
@@ -0,0 +1,91 @@
+import datetime
+import requests
+import json
+import time
+
+all_data = []
+project_list = []
+delta = datetime.timedelta(days=199)
+hour = datetime.timedelta(hours=1)
+zero = datetime.timedelta(hours=0)
+
+
+def save_json(start, end, out_path, input_path):
+    start_date = datetime.datetime.strptime(start, '%Y%m%d')
+    end_date = datetime.datetime.strptime(end, '%Y%m%d')
+    if (end_date - start_date).days < 0:
+        raise Exception('start date should NOT after end date')
+
+    expected_size = ((end_date - start_date).days + 1) * 24
+
+    # provide wiki-project list here!
+    with open(input_path) as f:
+        projects = f.read().splitlines()
+
+    for item in projects:
+        data = get_single_views(item, start_date, end_date)
+        if len(data['target']) == expected_size:
+            all_data.append(data)
+            project_list.append(item)
+
+    # convert all items to JSON and write to a file
+    print('saving data')
+    print(project_list)
+    with open('project_list.txt', 'w') as f:
+        for item in project_list:
+            f.write("%s\n" % item)
+    with open(out_path, 'w') as outfile:
+        outfile.write('\n'.join(json.dumps(i) for i in all_data) + '\n')
+    all_data.clear()
+    project_list.clear()
+
+
+def get_single_views(item, start_date, end_date):
+    base_url = 'https://wikimedia.org/api/rest_v1/metrics/pageviews/aggregate/' \
+               + item \
+               + '/all-access/user/hourly/'
+    # dict is used to save JSON of each domain

```

Comment:
generally you want all your comments to be of the form "<singlespace> Capitalized word in a sentence with a period." I relieze that you will still want to keep dict lower case as that is correct syntax. but its good ot have periods ending your comments.